### PR TITLE
Vue3: replace `baseParse` with `parse`

### LINKED ISSUE #25523: [Bug]:  decodeEntities option is required in browser builds
### Describe the bug

While stories render fine, the following error is thrown in the dev console:

```plain
error Error: [@vue/compiler-core] decodeEntities option is required in browser builds.
    at baseParse (compiler-core.esm-bundler.js:2586:1)
    at getComponents (entry-preview-docs.mjs:8:1)
    at getTemplateComponents (entry-preview-docs.mjs:8:1)
    at generateSource (entry-preview-docs.mjs:8:1)
    at immediate (entry-preview-docs.mjs:8:1)
    at callWithErrorHandling (runtime-core.esm-bundler.js:158:1)
    at callWithAsyncErrorHandling (runtime-core.esm-bundler.js:166:1)
    at job (runtime-core.esm-bundler.js:1827:1)
    at doWatch (runtime-core.esm-bundler.js:1858:1)
    at watch (runtime-core.esm-bundler.js:1716:1)#
```

### To Reproduce

not sure I can provide this… :grimacing: 

### System

```bash
Storybook Environment Info:

  System:
    OS: Linux 6.2 Debian GNU/Linux 12 (bookworm) 12 (bookworm)
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    Shell: 5.2.15 - /bin/bash
  Binaries:
    Node: 18.19.0 - /tmp/xfs-38cf59af/node
    Yarn: 4.0.2 - /tmp/xfs-38cf59af/yarn <----- active
    npm: 10.2.3 - /usr/local/bin/npm
  npmPackages:
    @storybook/addon-actions: 7.6.7 => 7.6.7 
    @storybook/addon-essentials: 7.6.7 => 7.6.7 
    @storybook/addon-links: 7.6.7 => 7.6.7 
    @storybook/vue3: 7.6.7 => 7.6.7 
    @storybook/vue3-webpack5: 7.6.7 => 7.6.7 
    eslint-plugin-storybook: 0.6.15 => 0.6.15 
    storybook: 7.6.7 => 7.6.7
```

### Additional context

Might be related to / duplicate of #25496.

Searching for `decodeEntities` in the Storybook documentation yields no results:

storybook js org_

### PR DESCRIPTION
**Pull request by Clarkkkk**
Closes #25523

Currently if use render function and template in the story, the `@vue/compiler-core` shows an error:  `decodeEntities option is required in browser builds.`

## What I did

replace `baseParse` with `parse` from the `@vue/compiler-dom`, which uses an internal method for the `decodeEntities` in the default option.

```ts
// vuejs/core/packages/compiler-dom/src/index.ts
export function parse(template: string, options: ParserOptions = {}): RootNode {
  return baseParse(template, extend({}, parserOptions, options))
}

// vuejs/core/packages/compiler-dom/src/parserOptions.ts
export const parserOptions: ParserOptions = {
  parseMode: 'html',
  isVoidTag,
  isNativeTag: tag => isHTMLTag(tag) || isSVGTag(tag) || isMathMLTag(tag),
  isPreTag: tag => tag === 'pre',
  decodeEntities: __BROWSER__ ? decodeHtmlBrowser : undefined,
  // ...
}
```

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

**Comment by JReinhold** [comment]
A couple of questions:

1. Will this only support Vue 3.4 and up? Vue is a peer dependnecy, so users could have 3.0 that we still need to support.
2. You've changed the import from `'@vue/compiler-core'` to `'@vue/compiler-dom'` without adding it as a dependency. I don't think that works?

**Comment by Clarkkkk** [comment]
Sorry for the oversight. Vue 3.4 has a pretty big refactor indeed. I should be more careful and will do more tests to ensure the compatibility.

**Comment by Clarkkkk** [comment]
In further investigations, i find that `baseParse` causes errors since vue 3.2. The fix works in vue 3.2 as well. And it should work in vue 3.0 too since the code of `parse` and `baseParse` is the same between the two versions.

I also run some tests on over 100 templates extracted from [storyblok-design-system](https://github.com/storyblok/storyblok-design-system). Except for the templates that cause errors in `baseParse`, the output of `parse` and `baseParse` is exactly the same. The tests can be found [here](https://github.com/Clarkkkk/storybook/commit/056578675f4efd053362f6e5be2e1a0bbf9cd1c5)

**Comment by SavkaTaras** [comment]
We have similar issue. We use vue3-vite storybook and when we have custom template:

render: (args: any) => ({
  components: { Cmp },
  setup() {
    return { args };
  },
  template: `
    <div class="additional_class"><cmp></cmp></div>
  `
}), 

This triggers the following error:

error Error: [@vue/compiler-core] decodeEntities option is required in browser builds.
    at baseParse (chunk-T5TZU2RZ.js?v=65634b0b:13474:13)
    at getComponents (@storybook_vue3_dist_entry-preview-docs__mjs.js?v=65634b0b:5607:26)
    at getTemplateComponents (@storybook_vue3_dist_entry-preview-docs__mjs.js?v=65634b0b:5600:23)
    at generateSource (@storybook_vue3_dist_entry-preview-docs__mjs.js?v=65634b0b:5641:143)
    at watch.immediate (@storybook_vue3_dist_entry-preview-docs__mjs.js?v=65634b0b:5636:13)
    at callWithErrorHandling (chunk-T5TZU2RZ.js?v=65634b0b:1861:18)
    at callWithAsyncErrorHandling (chunk-T5TZU2RZ.js?v=65634b0b:1869:17)
    at job (chunk-T5TZU2RZ.js?v=65634b0b:3608:9)
    at doWatch (chunk-T5TZU2RZ.js?v=65634b0b:3646:7)
    at watch (chunk-T5TZU2RZ.js?v=65634b0b:3473:10)


This is only in development mode. Prod doesn't have this error.

Storybook 7.6.7.

Thank you.

File: code/renderers/vue3/package.json [hunk 1]
```diff
@@ -62,7 +63,7 @@
     "@types/prettier": "2.7.2",
     "@vitejs/plugin-vue": "^4.4.0",
     "typescript": "^5.3.2",
-    "vue": "^3.2.47",
+    "vue": "^3.4.5",
```
Comment (chakAs3) [inline code/renderers/vue3/package.json:66]: this change has not impact as vue is peerDeps and the latest patch will be installed anyway

File: code/renderers/vue3/src/docs/sourceDecorator.ts [hunk 2]
```diff
@@ -13,7 +13,7 @@ import type {
   InterpolationNode,
   TemplateChildNode,
 } from '@vue/compiler-core';
-import { baseParse } from '@vue/compiler-core';
+import { parse } from '@vue/compiler-dom';
```
Comment (chakAs3) [inline code/renderers/vue3/src/docs/sourceDecorator.ts:16]: i don't need the entire compiler-dom for this feature,  compiler-core is minimum  ( baseParse ) i don't see why you need dom ?

**Review by chakAs3 — COMMENTED** [review]
i suggest to close this PR, and create a repoduction repo for the issue so i can help find right fix. Thanks @Clarkkkk

**Comment by Clarkkkk** [comment]
Thank you, it would be better for you to fix it. Here is the [reproduction repo](https://github.com/Clarkkkk/storyblok-design-system).

**Pull request closed without merging**